solver/src/solver/matcha_solver.rs

```diff
@@ -0,0 +1,312 @@
+        ensure!(
+            chain_id == MAINNET_CHAIN_ID,
+            "Matcha solver only supported on Mainnet",
+        );
+
+        let allowance_fetcher = settlement_contract.clone();
+        Ok(Self {
+            settlement_contract,
+            allowance_fetcher,
+            client: Box::new(DefaultMatchaApi::default()),
+        })
+    }
+}
+impl<F> MatchaSolver<F> {
+    fn web3(&self) -> DynWeb3 {
+        self.settlement_contract.raw_instance().web3()
+    }
+}
+
+impl<F> std::fmt::Debug for MatchaSolver<F> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("MatchaSolver")
+    }
+}
+
+#[async_trait::async_trait]
+impl<F: AllowanceFetching> SingleOrderSolving for MatchaSolver<F> {
+    async fn settle_order(&self, order: LimitOrder) -> Result<Option<Settlement>> {
+        let (swap, mut settlement) = match order.kind {
+            OrderKind::Sell => {
+                let query = SwapQuery {
+                    sell_token: order.sell_token,
+                    buy_token: order.buy_token,
+                    sell_amount: Some(order.sell_amount),
+                    buy_amount: None,
+                    slippage_percentage: Slippage::basis_points(STANDARD_MATCHA_SLIPPAGE_BPS)
+                        .unwrap(),
+                    skip_validation: Some(true),
+                };
+
+                tracing::debug!("querying Matcha swap api with {:?}", query);
+                let swap = self.client.get_swap(query).await?;
+                tracing::debug!("proposed Matcha swap is {:?}", swap);
+
+                ensure!(
+                    swap.buy_amount >= order.buy_amount,
+                    "order limit price not respected",
+                );
+                let settlement = Settlement::new(hashmap! {
+                    order.sell_token => swap.sell_amount,
+                    order.buy_token => swap.buy_amount,
+                });
+                (swap, settlement)
+            }
+            OrderKind::Buy => {
+                let query = SwapQuery {
+                    sell_token: order.sell_token,
+                    buy_token: order.buy_token,
+                    sell_amount: None,
+                    buy_amount: Some(order.buy_amount),
+                    slippage_percentage: Slippage::basis_points(STANDARD_MATCHA_SLIPPAGE_BPS)
+                        .unwrap(),
+                    // From the api documentation:
+                    // SlippagePercentage(Optional): The maximum acceptable slippage in % of the buyToken amount if sellAmount is provided, the maximum acceptable slippage in % of the sellAmount amount if buyAmount is provided. This parameter will change over time with market conditions.
+                    // => Hence, allegedly, we don't need to build in buffers ourselves. Though the sell amount is not adjusted, if slippage is changed for buy order requests.
+                    // Todo: Continue discussion with the 0x-team
```

Comment:
Do you mean the matcha team? I don't see how 0x is relevant here.

Reply:
0x and matcha are the same teams. matcha is developed by 0x :)